File: files/2026-04-19/CSEC 19042026.R
library(tidyverse)
library(readxl)

path <- "2026-04-19/Challenge 113.xlsx"
input <- read_excel(path, range = "B2:C7")
test <- read_excel(path, range = "E2:F5")

result = input %>%
  mutate(Date = as.Date(Date)) %>%
  group_by(Customer) %>%
  complete(Date = seq.Date(min(Date), max(Date), by = "day")) %>%
  filter(!Date %in% as.Date(input$Date)) %>%
  mutate(Date = as.POSIXct(Date))

all.equal(result, test, check.attributes = FALSE)
#> [1] TRUE

File: files/2026-04-19/CSEC 19042026.py
import pandas as pd

path = "2026-04-19/Challenge 113.xlsx"
input = pd.read_excel(path, usecols="B:C", nrows=6, skiprows=1)
test = pd.read_excel(path, usecols="E:F", nrows=3, skiprows=1)

result = (
    input.groupby("Customer")
      .apply(lambda x: pd.DataFrame({
          "Date": pd.date_range(x["Date"].min(), x["Date"].max())
      }).assign(Customer=x.name)
      .merge(x := input, on=["Customer","Date"], how="left", indicator=True)
      .query("_merge == 'left_only'")
      .drop(columns="_merge")
      .reset_index(drop=True)
))
result = result[["Customer", "Date"]].rename(columns={"Date": "Missing Date", "Customer": "Customer.1"}).reset_index(drop=True)

print(result.equals(test))
# True